sdk/java/pulumi/src/main/java/com/pulumi/deployment/internal/DeploymentImpl.java

```diff
@@ -1700,43 +1699,31 @@ public DefaultRunner(Logger standardLogger, EngineLogger engineLogger) {
             this.engineLogger = Objects.requireNonNull(engineLogger);
         }
 
-        public List<Exception> getSwallowedExceptions() {
-            return ImmutableList.copyOf(this.swallowedExceptions);
-        }
-
-        @Override
-        public <T extends Stack> CompletableFuture<Integer> runAsync(Supplier<T> stackFactory) {
-            try {
-                var stack = stackFactory.get();
-                var stackInternal = Internal.from(stack);
-                // Stack doesn't call RegisterOutputs, so we register them on its behalf.
-                stackInternal.registerPropertyOutputs();
-                registerTask(String.format("runAsync: %s, %s", stack.getResourceType(), stack.getResourceName()),
-                        Internal.of(stackInternal.getOutputs()).getDataAsync());
-            } catch (Exception ex) {
-                return handleExceptionAsync(ex);
-            }
-
-            return whileRunningAsync();
-        }
-
-        @Override
-        public CompletableFuture<Integer> runAsyncFuture(Supplier<CompletableFuture<Map<String, Output<?>>>> callback) {
-            return runAsyncFuture(callback, StackOptions.Empty);
-        }
-
         @Override
-        public CompletableFuture<Integer> runAsyncFuture(Supplier<CompletableFuture<Map<String, Output<?>>>> callback, StackOptions options) {
-            try {
-                var stack = StackInternal.of(callback, options);
-                // no outputs to register here
-                registerTask(String.format("runAsyncFuture: %s, %s", stack.getResourceType(), stack.getResourceName()),
-                        Internal.of(Internal.from(stack).getOutputs()).getDataAsync());
-            } catch (Exception ex) {
-                return handleExceptionAsync(ex);
-            }
-
-            return whileRunningAsync();
+        public <T> CompletableFuture<Result<T>> runAsync(Supplier<T> callback) {
+            var valueFuture = CompletableFuture.supplyAsync(callback);
+            // run the callback asynchronously in the context of the error handler
+            registerTask("DefaultRunner#runAsync", valueFuture);
+            // loop starts after the callback
+            return valueFuture
+                    .thenCompose(__ -> whileRunningAsync())
+                    .handle((code, throwable) -> {
```

Comment:
I wanted to make sure we always handle exceptions.

Reply:
Doesn't registerTask handle exceptions also? 

This might be ok but a bit convoluted to reason about.

Reply:
Let me check where runAsync is called.

Reply:
OK. This implements the project entry point. I think this isn't right. This code makes the `CompletableFuture<Result<T>>` ALWAYS wait for `valueFuture` which removes the feature of early termination due to an out-of-band exception somewhere in the underlying net of futures. That is, our protocol up to now, was that if any exception happens anywhere in the Output chain, we bubble up and stop. This is the point of registerTask and whileRunningAsync.

Is this an intentional change to the semantics?

I wonder what happened with the test that tested for it.

Reply:
> Is this an intentional change to the semantics?
No, it is not intentional. I was under the impression, this maintains the semantics, but I'm not sure anymore.

> I wonder what happened with the test that tested for it.
I'd guess it either does not test this conditions well enough or the semantics didn't actually changed.

I was under the impression that `valueFuture` would short circuit on an exception and not even call the `.thenCompose(__ -> whileRunningAsync())` and go straight to `handle`.

Let me try to come up with a test for those short-circuiting semantincs.

Reply:
After adding more test cases I was not successful in triggering the `if (throwable != null)` branch.

It turns out it might be redundant with the code in the loop itself. I wonder if we could remove the handler form the loop and handle exceptions here instead maintaining the semantics.

Reply:
Yes, after moving the handler to only be present in `handle` all tests pass and it looks equivalent (in behaviour) to me looking at the code.